From b8ef1526e2382bb96cd8643962a7176140d7b9b8 Mon Sep 17 00:00:00 2001
From: Jan Beulich <jbeulich@suse.com>
Date: Wed, 9 Jul 2025 10:59:42 +0200
Subject: x86/mwait-idle: adjust section annotations

idle_cpu_spr was bogusly non-const. Several other objects can be changed
from __read_mostly to __ro_after_init, for being altered from __init
functions only.

Signed-off-by: Jan Beulich <jbeulich@suse.com>
Acked-by: Andrew Cooper <andrew.cooper3@citrix.com>

diff --git a/xen/arch/x86/cpu/mwait-idle.c b/xen/arch/x86/cpu/mwait-idle.c
index ad421d8bb76a..4d24c83ed3ba 100644
--- a/xen/arch/x86/cpu/mwait-idle.c
+++ b/xen/arch/x86/cpu/mwait-idle.c
@@ -117,14 +117,14 @@ struct idle_cpu {
 	enum c1e_promotion c1e_promotion;
 };
 
-static const struct idle_cpu *icpu;
+static const struct idle_cpu *__ro_after_init icpu;
 
 static const struct cpuidle_state {
 	char		name[16];
 	unsigned int	flags;
 	unsigned int	exit_latency; /* in US */
 	unsigned int	target_residency; /* in US */
-} *cpuidle_state_table;
+} *__ro_after_init cpuidle_state_table;
 
 #define CPUIDLE_FLAG_DISABLED		0x1
 /*
@@ -515,7 +515,7 @@ static const struct cpuidle_state bdw_cstates[] = {
 	{}
 };
 
-static struct cpuidle_state __read_mostly skl_cstates[] = {
+static struct cpuidle_state __ro_after_init skl_cstates[] = {
 	{
 		.name = "C1",
 		.flags = MWAIT2flg(0x00),
@@ -567,7 +567,7 @@ static struct cpuidle_state __read_mostly skl_cstates[] = {
 	{}
 };
 
-static struct cpuidle_state __read_mostly skx_cstates[] = {
+static struct cpuidle_state __ro_after_init skx_cstates[] = {
 	{
 		.name = "C1",
 		.flags = MWAIT2flg(0x00) | CPUIDLE_FLAG_IRQ_ENABLE,
@@ -621,7 +621,7 @@ static const struct cpuidle_state icx_cstates[] = {
  * By default we enable C1E and disable C1 by marking it with
  * 'CPUIDLE_FLAG_DISABLED'.
  */
-static struct cpuidle_state __read_mostly adl_cstates[] = {
+static struct cpuidle_state __ro_after_init adl_cstates[] = {
 	{
 		.name = "C1",
 		.flags = MWAIT2flg(0x00) | CPUIDLE_FLAG_DISABLED,
@@ -655,7 +655,7 @@ static struct cpuidle_state __read_mostly adl_cstates[] = {
 	{}
 };
 
-static struct cpuidle_state __read_mostly adl_l_cstates[] = {
+static struct cpuidle_state __ro_after_init adl_l_cstates[] = {
 	{
 		.name = "C1",
 		.flags = MWAIT2flg(0x00) | CPUIDLE_FLAG_DISABLED,
@@ -689,7 +689,7 @@ static struct cpuidle_state __read_mostly adl_l_cstates[] = {
 	{}
 };
 
-static struct cpuidle_state __read_mostly spr_cstates[] = {
+static struct cpuidle_state __ro_after_init spr_cstates[] = {
 	{
 		.name = "C1",
 		.flags = MWAIT2flg(0x00),
@@ -805,7 +805,7 @@ static const struct cpuidle_state knl_cstates[] = {
 	{}
 };
 
-static struct cpuidle_state __read_mostly bxt_cstates[] = {
+static struct cpuidle_state __ro_after_init bxt_cstates[] = {
 	{
 		.name = "C1",
 		.flags = MWAIT2flg(0x00),
@@ -1102,15 +1102,15 @@ static const struct idle_cpu idle_cpu_icx = {
 	.c1e_promotion = C1E_PROMOTION_DISABLE,
 };
 
-static struct idle_cpu __read_mostly idle_cpu_adl = {
+static struct idle_cpu __ro_after_init idle_cpu_adl = {
 	.state_table = adl_cstates,
 };
 
-static struct idle_cpu __read_mostly idle_cpu_adl_l = {
+static struct idle_cpu __ro_after_init idle_cpu_adl_l = {
 	.state_table = adl_l_cstates,
 };
 
-static struct idle_cpu __read_mostly idle_cpu_spr = {
+static const struct idle_cpu idle_cpu_spr = {
 	.state_table = spr_cstates,
 	.c1e_promotion = C1E_PROMOTION_DISABLE,
 };
